perf/src/sigverify.rs

```diff
@@ -420,34 +423,76 @@ pub fn generate_offsets(
     )
 }
 
-fn dedup_packet(count: &AtomicU64, packet: &mut Packet, bloom: &AtomicBloom<&[u8]>) {
-    // If this packet was already marked as discard, drop it
-    if packet.meta.discard() {
-        return;
+pub struct Deduper {
+    filter: Vec<AtomicU64>,
+    seed: (u128, u128),
+    age: Instant,
+    max_age: Duration,
+    pub saturated: AtomicBool,
+}
+
+impl Deduper {
+    pub fn new(size: u32, max_age_ms: u64) -> Self {
+        let mut filter: Vec<AtomicU64> = Vec::with_capacity(size as usize);
+        filter.resize_with(size as usize, Default::default);
+        let seed = thread_rng().gen();
+        Self {
+            filter,
+            seed,
+            age: Instant::now(),
+            max_age: Duration::from_millis(max_age_ms),
+            saturated: AtomicBool::new(false),
+        }
     }
 
-    // If this packet was not newly added, it's a dup and should be discarded
-    if !bloom.add(&&packet.data.as_slice()[0..packet.meta.size]) {
-        packet.meta.set_discard(true);
-        count.fetch_add(1, Ordering::Relaxed);
+    pub fn reset(&mut self) {
+        let now = Instant::now();
+        //this should reset every 500k unique packets per 1m sized deduper
+        //false positive rate is 1/1000 at that point
+        let saturated = self.saturated.load(Ordering::Relaxed);
+        if saturated || now.duration_since(self.age) > self.max_age {
+            for i in &self.filter {
+                i.store(0, Ordering::Relaxed);
+            }
+            self.seed = thread_rng().gen();
+            self.age = now;
+            self.saturated.store(false, Ordering::Relaxed);
```

Comment:
It looks like there's ordering and visibility assumptions in these atomics. Specifically around `.saturated.store()` and `.filter[pos].store()`. Since these are both relaxed, they can be reordered w.r.t. each other. Since the PR/code mentions this'll be running in parallel, it'd be possible for a threads to see `.saturated.store(false)` before the filters are cleared.

If that is correct, then I think these orderings should be bumped up, such that the loads become Acquire and the stores become Release. On x86 this is basically free. On arm it becomes correct 😅.

This would also need to apply to all the atomic load/stores of `saturated` and `filter`. With `self.filter[pos].fetch_or()` becoming `AcqRel`. Since the filter is checked before `saturated`, it's the same reordering/visibility issue.

Benchmarking locally on my intel x86 MBP, I saw basically the same performance.

Baseline
```
test bench_dedup_baseline           ... bench:          43 ns/iter (+/- 2)
test bench_dedup_diff_big_packets   ... bench:     416,778 ns/iter (+/- 75,811)
test bench_dedup_diff_small_packets ... bench:     155,807 ns/iter (+/- 19,780)
test bench_dedup_reset              ... bench:     314,559 ns/iter (+/- 33,493)
test bench_dedup_same_big_packets   ... bench:     376,994 ns/iter (+/- 166,838)
test bench_dedup_same_small_packets ... bench:     116,076 ns/iter (+/- 10,169)
```

Acquire/Release on .saturated
```
test bench_dedup_baseline           ... bench:          43 ns/iter (+/- 2)
test bench_dedup_diff_big_packets   ... bench:     423,094 ns/iter (+/- 76,294)
test bench_dedup_diff_small_packets ... bench:     154,026 ns/iter (+/- 25,957)
test bench_dedup_reset              ... bench:     313,213 ns/iter (+/- 24,235)
test bench_dedup_same_big_packets   ... bench:     363,302 ns/iter (+/- 164,439)
test bench_dedup_same_small_packets ... bench:     114,966 ns/iter (+/- 24,617)
```

Commit here: https://github.com/solana-labs/solana/commit/813e526387e8ab9d4273ced8d237465c8dbcd1df

Reply:
ordering should make a difference on how this functions since its tolerant of a few false positives or a few false negatives.  bad behavior would be if it gets stuck in a loop constantly resetting the whole filter.  but I don't think that would be the case.